File: DaliyTemperatures.py
#stack approach

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        answer = [0] * len(temperatures) #initialize all answer as 0
        
        for i, cur in enumerate(temperatures):
            while stack and cur > temperatures[stack[-1]]: #if current temperature is higher than the last one.
                last = stack.pop()                         #extract the last index
                answer[last] = i - last                    #add the difference between current index and 
            
            stack.append(i)
            
        return answer 

File: DaliyTemperatures.java
class Solution {
    
    public int[] dailyTemperatures(int[] temperature){
        
        Stack<int[]> st = new Stack();
        int [] res = new int[temperature.length];
        
        for(int i=0;i<temperature.length;i++){
            
            while(!st.isEmpty() && st.peek()[0] < temperature[i]){
                int [] prevTemp = st.pop();
                res[prevTemp[1]] = i - prevTemp[1];
            }
            
            int[] curTemp = new int[]{temperature[i], i};
            st.push(curTemp);
            
        }
        
        return res;
        
    }
    
    
//     static class Temp{
        
//         public int value;
//         public int index;
        
//         public Temp(int v, int i){
//             this.value = v;
//             this.index = i;
//         }
        
//     }
    
    
//     public int[] dailyTemperatures(int[] temperatures) {
        
//         Stack<Temp> st = new Stack<>();
//         int [] res = new int[temperatures.length];
    
//         for(int i=0;i<temperatures.length;i++){
            
//             while(!st.isEmpty() && st.peek().value < temperatures[i]){
                    
//                     Temp t = st.pop();
//                     res[t.index] = i - t.index;
                
//                 }
//             st.push(new Temp(temperatures[i], i));
//         }
        
//         return res;
//     }
}
